README.md
# clojureverse

A sandbox to tinker with the Clojure universe.

## Installation

This project uses [Leiningen](https://leiningen.org/) as the package manager.

## Usage

To install dependencies

```
lein deps
```

To execute, simply

```
lein run
```

To run tests

```
lein test
```

## Resources

- [Clojure](https://clojure.org/)
- [Community Docs](https://clojuredocs.org/)
- [Calva.io](https://calva.io/)
- [Clojars](https://clojars.org/)
- [Clojure programming world: 6 essential libraries and other tools](https://freshcodeit.com/freshcode-post/clojure-programming-world-frameworks-and-other-tools)


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2023-05-28
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2023-05-28
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://sourcehost.site/your-name/clojureverse/compare/0.1.1...HEAD
[0.1.1]: https://sourcehost.site/your-name/clojureverse/compare/0.1.0...0.1.1


test/concepts/flow_test.clj
(ns concepts.flow-test
  (:require [clojure.test :refer :all]
            [concepts.flow :refer :all]))


(deftest odd-or-even-test
  (testing "When number is 3 return odd"
    (is (= (odd-or-even 3) "odd")))
  
  (testing "When number is 4 return even"
    (is (= (odd-or-even 4) "even"))))

(deftest odd-or-even-side-effect-test
  (testing "When number is 3 return odd"
    (is (= (odd-or-even-side-effect 3) "odd")))
  
  (testing "When number is 4 return even"
    (is (= (odd-or-even-side-effect 4) "even"))))

(deftest greet-test
  (testing "When name is given should greet",
    (is (= (greet "Alice") "hello Alice")))
  
  (testing "When no name is provided return nothing"
    (is (= (greet nil) nil))))

(deftest status-test
  (testing "Return OK when 200"
    (is (= (status 200) "OK")))
  
  (testing "Return Bad Request when 400"
    (is (= (status 400) "Bad Request")))
  
  (testing "Return Not Found when 404"
    (is (= (status 404) "Not Found")))
  
  (testing "Return Unknown when not supported"
    (is (= (status 123) "Unknown"))))

(deftest temperature-test
  (testing "Return Extremely Hot when temperature >= 120"
    (is (= (temperature 121) "Extremely Hot")))
  
  (testing "Return Hot when temperature >= 100 < 120"
    (is (= (temperature 101) "Hot")))
  
  (testing "Return Mild when temperature >= 80 < 100"
    (is (= (temperature 81) "Mild")))
  
  (testing "Return Cool when temperature >= 60 < 80"
    (is (= (temperature 61) "Cool")))
  
  (testing "Return Cold when temperature >= 0 < 60"
    (is (= (temperature 1) "Cold")))
  
  (testing "Return Frigid when temperature below 0"
    (is (= (temperature -1) "Frigid"))))



test/clojureverse/core_test.clj
(ns clojureverse.core-test
  (:require [clojure.test :refer :all]
            [clojureverse.core :refer :all]
            [cheshire.core :as cheshire]))

(deftest a-test
  (testing "I Success."
    (is (= 1 1))))

;; Usage of cheshire/generate-string
(deftest cheshire-generate-string-test
  (testing "Stringify map with 2 properties"
    (def person {:name "Sandra" :age 31})
    (is (= (cheshire/generate-string person) "{\"name\":\"Sandra\",\"age\":31}"))))


test/concepts/functions_test.clj
(ns concepts.functions-test
  (:require [clojure.test :refer :all]
            [concepts.functions :refer :all]))

(deftest prefix-strs-test
  (testing "Passed in the argument, gets prefixed with `prefix-`"
    (is (= (prefix-strs "hello") "prefix-hello"))))

(deftest filter-map-test
  (testing "Given an array of maps, get only the maps that have :value = 3"
    (def my-map [{:name "Alice" :value 1} {:name "Bruce" :value 2} {:name "Camelia" :value 3}])
    (def target-value 3)
    (is (= (filter-maps-by-value my-map target-value) `({:name "Camelia" :value 3})))))

(deftest addition-test
  (testing "Test 1 + 2 = 3"
    (is (= (addition 1 2) 3))))

(deftest get-delimiter-test
  (testing "When no args is passed, expect `you have no args`"
    (is (= (get-delimiter) "you have no args")))

  (testing "When one arg is passed, expect args to be suffixed with `:`"
    (is (= (get-delimiter "hello") "hello:")))

  (testing "When two args is passed, expect args to be delimited with `-`"
    (is (= (get-delimiter "hello" "world") "hello-world")))

  (testing "When three args is passed, expect args to be delimited with `;`"
    (is (= (get-delimiter "hello" "hello" "hello") "hello;hello;hello"))))

(deftest aggregate-counts-test
  (testing "When I pass in [1 2 3], expect sum to be 6"
    (is (= (aggregate-counts 1 2 3) 6)))

  (testing "When I pass in [10 20 30 40 60], expect sum to be 160"
    (is (= (aggregate-counts 10, 20, 30, 40, 60) 160))))


(deftest ackermann-test
  (testing "Ackermann(1, 2) = 4"
    (is (= (ackermann 1 2) 4)))

  (testing "Ackermann(2, 1) = 5"
    (is (= (ackermann 2 1) 5)))

  (testing "Ackermann(2, 2) = 7"
    (is (= (ackermann 2 2) 7))))


(deftest factorial-test
  (testing "6! = 720"
    (is (= (factorial 6) 720))))


(deftest people-in-scenes-test
  (def scenes [{:subject "Frankie"
                :action "say"
                :object "relax"}

               {:subject "Lucy"
                :action "❤s"
                :object "Clojure"}

               {:subject "Rich"
                :action "tries"
                :object "a new conditioner"}])

  (testing "Generate a string with the names of people in scene"
    (is (= (people-in-scenes scenes) "Frankie, Lucy, Rich"))))


src/concepts/dtypes.clj
(ns concepts.dtypes)


;; Strings
(println "This is a string")

; Defining a variable
(def my-str-var (str "This is", " a string"))
(println my-str-var)
(println (type my-str-var))

;; Characters
(def my-char \Z)
(println my-char)
(println (type my-char))

;; Regular Expressions
(def my-regex #"Clojure")
(println (type my-regex))

; Finding match
(println(re-find my-regex "Clojure is great!"))
(println(re-find my-regex "Scala is great!"))

;; Ints
(def my-int 100)
(println (type my-int))

;; Floats
(def my-decimal 3.14159)
(println (type my-decimal))

; Addition of int and float
(def result (+ my-int my-decimal))
(println result)
(println (type result))

;; Ratio
(def my-ratio 100/8)
(def prefix-ratio (/ 2 3))
(def whole-number-ratio 10/5)

(println (type my-ratio))
(println (type prefix-ratio))
(println (type whole-number-ratio))

; Gets denominator
(println (denominator my-ratio)) ; where 100/8 becomes 25/2

; Check is ratio
(println (ratio? my-ratio)) ; the function that ends with ? indicates a boolean returned

;; Vectors
(def my-vector [1 2 3 4])
(println (filter even? my-vector))

;; List
(def my-list `(1 2 3 4))
(println (filter odd? my-list))

;; Sets
(def my-set #{1 2 3 4 5}) ; Set is unordered. Error if duplicates found
(println my-set)

(def converted-to-set (set [1 2 1 2 3 4 2 3 1 1]))
(println converted-to-set)

;; Maps
(def my-map {:a 1 :b 2 :c 3})

; Merge maps
(def my-map-2 {:a 4 :d 4 :e 6})
(println (merge my-map my-map-2))

; Print values of map
(println (vals my-map))

;; Symbol
(println (type map))
(println (map (fn [num] (+ num 1)) `(1, 2, 3, 4, 5)))
(println (type nil))
(println (type true))
(println (type false))

:my-keyword
(println (type :my-keyword))

;; deftype (Programming abstractions)
; generates a new type with a constructor
(deftype Position [x y])

;; defrecord
; generates an immutable, persistent map
(defrecord Person [name address])


src/concepts/functions.clj
(ns concepts.functions)

;; Regular Functions
(defn prefix-strs
  [str-to-be-prefixed]
  (str "prefix-" str-to-be-prefixed))

;; Anonymous function
(defn filter-maps-by-value [maps value]
  (filter #(= value (:value %)) maps))

;; Assign anonymous function to a variable
(def addition #(+ %1 %2))


;; Multi-arity function
(defn get-delimiter
  ([] "you have no args")
  ([first-arg] (str first-arg ":"))
  ([first-arg, second-arg] (str first-arg "-" second-arg))
  ([first-arg, second-arg, third-arg] (str first-arg ";" second-arg ";" third-arg)))

;; Variadic function
(defn aggregate-counts [& counts]
  (apply + counts))

;; Recursion
(defn ackermann [m n]
  (case m
    0 (inc n)
    (case n
      0 (ackermann (dec m) 1)
      (ackermann (dec m) (ackermann m (dec n))))))

(defn factorial [n]
  (if (zero? n)
    1
    (* n (factorial (dec n)))))

;; Piping functions
(defn people-in-scenes [scenes]
  (->> scenes
       (map :subject)
       (interpose ", ")
       (reduce str)))


src/concepts/flow.clj
(ns concepts.flow)

;; If statement
(defn odd-or-even [number]
  (if (odd? number) "odd" "even"))

(defn odd-or-even-side-effect [number]
  (if (odd? number)
    (do (println "this is odd") "odd")
    (do (println "this is even") "even")))

(defn greet [name]
  (when name
    (str "hello " name)))

(defn status [code]
  (case code
    200 "OK"
    400 "Bad Request"
    404 "Not Found"
    "Unknown"))

(defn temperature [temp]
  (cond
    (>= temp 120) "Extremely Hot"
    (>= temp 100) "Hot"
    (>= temp 80) "Mild"
    (>= temp 60) "Cool"
    (>= temp 0) "Cold"
    :else "Frigid"))

;; Iteration with side-effect
(defn iterate-print [count]
  (dotimes [current count]
    (println "Current: " current)))

(defn nested-iterate [arr1 arr2]
  (doseq [a arr1 b arr2]
    (println (* a b))))

;; Iteration without side-effect
(defn list-comprehension [arr1 arr2]
  (for [a arr1 b arr2]
    (* a b)))

;; Exceptions - use for JVM interop
(defn must-more-than-zero [num]
  (if (> num 0)
    (throw (Exception. "something happened"))
    num))

(defn try-catch-finally []
  (try
    (must-more-than-zero 3)
    (catch RuntimeException e "oh no" e)
    (finally (println "always runs")))
  )
